### PR TITLE
D3D: More debug information and break on error

### PR DESCRIPTION
**Pull request by mrgreywater**
This change only affects the debug build and should give a little more information when debugging the DirectX backend. It also adds a breakpoint when otherwise the graphics driver is likely to crash. The function GetDebugObjectName is added but not used, it's helpful when trying to figure out which objects actually cause a failure.

The commits will be squashed with an appropriate message if you think this is worth adding. It certainly helped me when debugging.

### CODE REVIEW

File: Source/Core/VideoBackends/D3D/D3DBase.cpp [hunk 1]
```diff
@@ -302,6 +303,28 @@ HRESULT Create(HWND wnd)
...
+			{
+				infoQueue->SetBreakOnSeverity(D3D11_MESSAGE_SEVERITY_CORRUPTION, true);
+				infoQueue->SetBreakOnSeverity(D3D11_MESSAGE_SEVERITY_ERROR, true);
+
+				D3D11_MESSAGE_ID hide[] =
+				{
+					D3D11_MESSAGE_ID_SETPRIVATEDATA_CHANGINGPARAMS
+				};
+
+				D3D11_INFO_QUEUE_FILTER filter;
```
Comment (lioncash) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: `D3D11_INFO_QUEUE_FILTER filter = {};`

Comment (mrgreywater) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: I can surely change that, just note that the whole function uses memset for all other DX structs. See line 268, 277, 281.

**Comment by degasus** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/VideoBackends/D3D/D3DBase.cpp [hunk 2]
```diff
@@ -302,6 +303,28 @@ HRESULT Create(HWND wnd)
 											supported_feature_levels, NUM_SUPPORTED_FEATURE_LEVELS,
 											D3D11_SDK_VERSION, &swap_chain_desc, &swapchain, &device,
 											&featlevel, &context);
+		// Debugbreak on D3D error
+		if (SUCCEEDED(hr) && SUCCEEDED(device->QueryInterface(__uuidof(ID3D11Debug), (void**)&debug)))
+		{
+			ID3D11InfoQueue* infoQueue = nullptr;
+			if (SUCCEEDED(debug->QueryInterface(__uuidof(ID3D11InfoQueue), (void**)&infoQueue)))
+			{
+				infoQueue->SetBreakOnSeverity(D3D11_MESSAGE_SEVERITY_CORRUPTION, true);
```
Comment (neobrain) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: Wouldn't this usually be enabled manually in the DX control panel?

File: Source/Core/VideoBackends/D3D/D3DBase.cpp [hunk 3]
```diff
@@ -375,6 +398,21 @@ void Close()
 
 	SAFE_RELEASE(context);
 	ULONG references = device->Release();
+
+#if defined(_DEBUG) || defined(DEBUGFAST)
+	if (debug)
+	{
+		--references; // we didn't release the debug interface yet
```
Comment (neobrain) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: Why didn't we? Where is it released instead? What does `SAFE_RELEASE(debug)` do, since apparently it's not the debug interface? Why is references decremented instead of just moving this block in front of the actual `device->Release` call?

This comment needs some elaboration, because things are still confusing.

File: Source/Core/VideoBackends/D3D/D3DBase.cpp (hunk 2 above)
Comment (mrgreywater) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: The DX control panel allows to override this setting, sure. But the same applies to the the Debug Creation Flag which is also enabled one line above.

This modifies just the default behaviour, since most of the time you are not specifically debugging Directx when crashes occur, and it would be nice to hit an int3 instead of crashing the driver.

File: Source/Core/VideoBackends/D3D/D3DBase.cpp (hunk 3 above)
Comment (mrgreywater) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: the debug interface is released at 412. The debug interface itself increases the refcount of the device, so i have to substract that.

I have to do the debug check after i release the device, because otherwise I don't know the refCount of the device. Also, debug->ReportLiveDeviceObjects(D3D11_RLDO_SUMMARY | D3D11_RLDO_DETAIL); would otherwise show another live reference.

**Comment by delroth** [comment]
@dolphin-emu-bot rebuild

That looks pretty cool. Granted, I don't know a lot about DX...

@Armada651 care to review this? :)

**Comment by mrgreywater** [comment]
@Armada651 The memset change was done since lioncash wanted it, the comment was changed because of neobrain. 
Fixing the D3D11 warning of the efb texture shader ressource in the debug build is not directly related, I can split it in another PR if it's required, or rearrange the commits, so it's easily distinguishable..

**Comment by Armada651** [comment]
I think it's alright, just wanted to confirm it was intentional.

File: Source/Core/VideoBackends/D3D/XFBEncoder.cpp [hunk 4]
```diff
@@ -361,8 +361,8 @@ void XFBEncoder::Encode(u8* dst, u32 width, u32 height, const EFBRectangle& srcR
 	D3D::context->Unmap(m_outStage, 0);
 
 	// Restore API
-
 	g_renderer->RestoreAPIState();
+	D3D::stateman->Apply(); // force unbind efb texture as shader resource
```
Comment (magumagu) [inline Source/Core/VideoBackends/D3D/XFBEncoder.cpp]: Move this into a separate PR?
